network/message/authorization.go

```diff
@@ -0,0 +1,375 @@
+package message
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/libp2p/message"
+	"github.com/onflow/flow-go/model/messages"
+	"github.com/onflow/flow-go/network/channels"
+)
+
+// MsgAuthConfig contains authorization information for a specific flow message. The authorization
+// is represented as a map from network channel -> list of all roles allowed to send the message on
+// the channel.
+type MsgAuthConfig struct {
+	// Name is the string representation of the message type.
+	Name string
+	// Type a func that returns a new instance of message type.
+	Type func() interface{}
+	// Config is the mapping of network channel to list of authorized flow roles.
+	Config map[channels.Channel]flow.RoleList
+}
+
+// IsAuthorized checks if the specified role is authorized to send the message on channel and
+// asserts that the message is authorized to be sent on channel.
+// Expected error returns during normal operations:
+//  * ErrUnauthorizedMessageOnChannel: if channel does not exist in message config
+//  * ErrUnauthorizedRole: if list of authorized roles for message config does not include role
+func (m MsgAuthConfig) IsAuthorized(role flow.Role, channel channels.Channel) error {
+	authorizedRoles, ok := m.Config[channel]
+	if !ok {
+		return ErrUnauthorizedMessageOnChannel
+	}
+
+	if !authorizedRoles.Contains(role) {
+		return ErrUnauthorizedRole
+	}
+
+	return nil
+}
+
+var (
+	ErrUnknownMsgType               = errors.New("could not get authorization config for unknown message type")
+	ErrUnauthorizedMessageOnChannel = errors.New("message is not authorized to be sent on channel")
+	ErrUnauthorizedRole             = errors.New("sender role not authorized to send message on channel")
+	AuthorizationConfigs            map[string]MsgAuthConfig
+)
+
+func initializeMessageAuthConfigsMap() {
+	AuthorizationConfigs = make(map[string]MsgAuthConfig)
+
+	// consensus
+	AuthorizationConfigs[BlockProposal] = MsgAuthConfig{
+		Name: BlockProposal,
+		Type: func() interface{} {
+			return new(messages.BlockProposal)
+		},
+		Config: map[channels.Channel]flow.RoleList{
+			channels.ConsensusCommittee: {flow.RoleConsensus},
+			channels.PushBlocks:         {flow.RoleConsensus}, // channel alias ReceiveBlocks = PushBlocks
+		},
+	}
+	AuthorizationConfigs[BlockVote] = MsgAuthConfig{
+		Name: BlockVote,
+		Type: func() interface{} {
+			return new(messages.BlockVote)
+		},
+		Config: map[channels.Channel]flow.RoleList{
+			channels.ConsensusCommittee: {flow.RoleConsensus},
+		},
+	}
+
+	// protocol state sync
+	AuthorizationConfigs[SyncRequest] = MsgAuthConfig{
+		Name: SyncRequest,
+		Type: func() interface{} {
+			return new(messages.SyncRequest)
+		},
+		Config: map[channels.Channel]flow.RoleList{
+			channels.SyncCommittee:     flow.Roles(),
+			channels.SyncClusterPrefix: flow.Roles(),
```

Comment:
for this and all of the other sync cluster configs, are all roles authorized, or just collection nodes?

Reply:
https://github.com/onflow/flow-go/pull/2648/commits/7e74be1ae9418926d1edd59cbf45b11edf8896bd